chainer/iterators/order_samplers.py

```diff
@@ -0,0 +1,57 @@
+import numpy
+
+
+class OrderSampler(object):
+
+    """Base class of all order samplers.
+
+    Every order sampler subclass has to provide a method
+    :meth:`__call__`.
+    This method is called by an iterator before a new epoch,
+    and it should return a new index order for the next epoch.
+
+    """
+
+    def __call__(self, current_order, current_position):
+        """Sample the next order.
+
+        Args:
+            current_order (numpy.ndarray): 1-D array of indices.
+                The length should be the same as the dataset to sample
+                data from.
+            current_position (int): The current position of an iterator.
+
+        Returns:
+            numpy.ndarray: 1-D array of indices. This is the order in which
+                examples are sampled from a dataset in the next epoch.
+
+        """
+        raise NotImplementedError
+
+
+class ShuffleOrderSampler(OrderSampler):
+
+    """Sampler that generates random orders.
+
+    This is expected to be used together with Chainer's iterators.
+    An order sampler is called by an iterator every epoch.
+
+    The two lines below create basically the same objects.
+
+    >>> chainer.iterators.MultiProcessIterator(dataset, 1, shuffle=True)
+    >>> chainer.iterators.MultiProcessIterator(
+    ...     dataset, 1, order_sampler=ShuffleOrderSampler())
+
+    """
+
+    def __init__(self, seed=None):
+        # Use a distinct RandomState in the thread
+        # for deterministic random number generation.
+        # To support 32-bit platform and numpy < 1.11,
+        # the seed is taken in a verbose manner.
+        if seed is None:
+            seed = numpy.random.randint(2**31)
+        self._random = numpy.random.RandomState(seed)
```

Comment:
Are you still able to train with shuffle using the global numpy rng? I think this changes the default behavior actually for users who used to fix their seeds and then train with shuffle. How about instead of the seed accept a RandomState, and then use the global state by default?

Reply:
I basically copied the existing code.
Am I missing something?

https://github.com/chainer/chainer/blob/master/chainer/iterators/multiprocess_iterator.py#L288

Reply:
Well, that is only the case for `MultiprocessIterator`.
I will fix the code.